File: android/feature/scripture/src/main/kotlin/com/nexussoft/verbum/feature/scripture/ScriptureFeature.kt
package com.nexussoft.verbum.feature.scripture

import com.nexussoft.verbum.clients.BibleClient
import com.nexussoft.verbum.clients.ClipboardClient
import com.nexussoft.verbum.clients.PreferencesClient
import com.nexussoft.verbum.common.arch.Effect
import com.nexussoft.verbum.common.arch.Reducer
import com.nexussoft.verbum.common.arch.combine
import com.nexussoft.verbum.common.arch.only
import com.nexussoft.verbum.common.arch.pullback
import com.nexussoft.verbum.common.arch.pullbackOptional
import com.nexussoft.verbum.common.arch.with
import com.nexussoft.verbum.models.PassageReference

/**
 * The reading destination: a chapter reader with the bookshelf beside it — a sheet over the
 * page in compact width, a column to the left in expanded width. Twin of iOS `ScriptureFeature`.
 */
object ScriptureFeature {
    data class State(
        val books: BookPickerFeature.State,
        val reader: ChapterReaderFeature.State,
        val settings: ReaderSettingsFeature.State? = null,
        /** Compact width only: the shelf presented over the page. */
        val isShelfPresented: Boolean = false,
    ) {
        companion object {
            fun initial(reference: PassageReference = PassageReference("John", 3), textScale: ReaderTextScale = ReaderTextScale.STANDARD): State {
                val reader = ChapterReaderFeature.State(reference, textScale)
                return State(books = BookPickerFeature.State(current = reader.reference), reader = reader)
            }
        }
    }

    sealed interface Action {
        data class Books(val action: BookPickerFeature.Action) : Action
        data class Reader(val action: ChapterReaderFeature.Action) : Action
        data class Settings(val action: ReaderSettingsFeature.Action) : Action
        data object TitleTapped : Action
        data object ShelfDismissed : Action
        data object SettingsButtonTapped : Action
        data object SettingsDismissed : Action
        data class Delegate(val delegate: DelegateAction) : Action
    }

    sealed interface DelegateAction {
        data class Listen(val reference: PassageReference) : DelegateAction
        data class Talk(val reference: PassageReference) : DelegateAction
        data class OpenContext(val reference: PassageReference) : DelegateAction
    }

    fun reducer(bibleClient: BibleClient, clipboard: ClipboardClient, preferences: PreferencesClient): Reducer<State, Action> = combine(
        BookPickerFeature.reducer.pullback(
            get = { it.books }, set = { s, c -> s.copy(books = c) },
            extractAction = { (it as? Action.Books)?.action }, embedAction = { Action.Books(it) },
        ),
        ChapterReaderFeature.reducer(bibleClient, clipboard, preferences).pullback(
            get = { it.reader }, set = { s, c -> s.copy(reader = c) },
            extractAction = { (it as? Action.Reader)?.action }, embedAction = { Action.Reader(it) },
        ),
        ReaderSettingsFeature.reducer(preferences).pullbackOptional(
            get = { it.settings }, set = { s, c -> s.copy(settings = c) },
            extractAction = { (it as? Action.Settings)?.action }, embedAction = { Action.Settings(it) },
        ),
        Reducer { state, action ->
            when (action) {
                Action.TitleTapped -> state.copy(isShelfPresented = true).only()
                Action.ShelfDismissed -> state.copy(isShelfPresented = false).only()
                Action.SettingsButtonTapped -> state.copy(settings = ReaderSettingsFeature.State(state.reader.textScale)).only()
                Action.SettingsDismissed -> state.copy(settings = null).only()
                is Action.Books -> when (val a = action.action) {
                    is BookPickerFeature.Action.Delegate -> when (val d = a.delegate) {
                        is BookPickerFeature.DelegateAction.ChapterSelected ->
                            state.copy(isShelfPresented = false).with(Effect.Send(Action.Reader(ChapterReaderFeature.Action.Go(d.reference))))
                    }
                    else -> state.only()
                }
                is Action.Reader -> {
                    val synced = state.copy(books = state.books.copy(current = state.reader.reference))
                    when (val delegate = (action.action as? ChapterReaderFeature.Action.Delegate)?.delegate) {
                        is ChapterReaderFeature.DelegateAction.Listen -> synced.with(Effect.Send(Action.Delegate(DelegateAction.Listen(delegate.reference))))
                        is ChapterReaderFeature.DelegateAction.Talk -> synced.with(Effect.Send(Action.Delegate(DelegateAction.Talk(delegate.reference))))
                        is ChapterReaderFeature.DelegateAction.OpenContext -> synced.with(Effect.Send(Action.Delegate(DelegateAction.OpenContext(delegate.reference))))
                        null -> synced.only()
                    }
                }
                is Action.Delegate -> state.only()
                is Action.Settings -> when (val a = action.action) {
                    is ReaderSettingsFeature.Action.TextScaleChanged ->
                        state.with(Effect.Send(Action.Reader(ChapterReaderFeature.Action.TextScaleChanged(a.scale))))
                }
            }
        },
    )
}

File: VerbumKit/Sources/Features/Scripture/ScriptureFeature.swift
import ComposableArchitecture
import Models

/// The reading destination: a chapter reader with the bookshelf beside it.
///
/// Where the shelf lives is size-class driven, never orientation driven: in
/// compact width it is a sheet over the page (as Books shows its table of
/// contents); in regular width — iPad, an unfolded phone — it sits in a column
/// to the left. Search and entity pages are the app shell's business now.
@Reducer
public struct ScriptureFeature {
    @ObservableState
    public struct State: Equatable {
        public var books: BookPickerFeature.State
        public var reader: ChapterReaderFeature.State
        @Presents public var settings: ReaderSettingsFeature.State?
        /// Compact width only: the shelf presented over the page.
        public var isShelfPresented = false

        public init(reference: PassageReference = PassageReference(bookId: "John", chapter: 3)) {
            let reader = ChapterReaderFeature.State(reference: reference)
            self.reader = reader
            self.books = BookPickerFeature.State(current: reader.reference)
        }
    }

    public enum Action: Equatable {
        case books(BookPickerFeature.Action)
        case reader(ChapterReaderFeature.Action)
        case settings(PresentationAction<ReaderSettingsFeature.Action>)
        case titleTapped
        case shelfDismissed
        case settingsButtonTapped
        case delegate(Delegate)

        @CasePathable
        public enum Delegate: Equatable {
            case listen(PassageReference)
            case talk(PassageReference)
            case openContext(PassageReference)
        }
    }

    public init() {}

    public var body: some ReducerOf<Self> {
        Scope(state: \.books, action: \.books) {
            BookPickerFeature()
        }
        Scope(state: \.reader, action: \.reader) {
            ChapterReaderFeature()
        }
        Reduce { state, action in
            switch action {
            case .titleTapped:
                state.isShelfPresented = true
                return .none

            case .shelfDismissed:
                state.isShelfPresented = false
                return .none

            case .settingsButtonTapped:
                state.settings = ReaderSettingsFeature.State()
                return .none

            case .books(.delegate(.chapterSelected(let reference))):
                state.isShelfPresented = false
                return .send(.reader(.go(to: reference)))

            case .reader(.delegate(.listen(let reference))):
                return .send(.delegate(.listen(reference)))

            case .reader(.delegate(.talk(let reference))):
                return .send(.delegate(.talk(reference)))

            case .reader(.delegate(.openContext(let reference))):
                return .send(.delegate(.openContext(reference)))

            case .reader:
                // Keep the shelf's "you are here" in step with the reader.
                state.books.current = state.reader.reference
                return .none

            case .books, .settings, .delegate:
                return .none
            }
        }
        .ifLet(\.$settings, action: \.settings) {
            ReaderSettingsFeature()
        }
    }
}
